### PR TITLE
Feature/21958 ess cache search or download functionality of enc files

### PR DESCRIPTION
**Pull request by Shirin14926**
caching search and download code changes and unit test cases

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                if (internalProductItemNotFound.UpdateNumbers != null && internalProductItemNotFound.UpdateNumbers.Any())
+                {
+                    internalProductsNotFound.Add(internalProductItemNotFound);
+                }
+            }
+
+            return internalProductsNotFound;
+        }
+
+        private async Task<BatchDetail> CheckIfCacheProductsExistsinBlob(string exchangeSetRootPath, SalesCatalogueServiceResponseQueueMessage queueMessage, Products item, List<int?> updateNumbers, int? itemUpdateNumber, string storageConnectionString, FssResponseCache cacheInfo)
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:97]: Kindly change casing of "in" to "In" in method name

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IAzureTableStorageClient.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+using System.Threading.Tasks;
+
+namespace UKHO.ExchangeSetService.Common.Helpers
+{
+    public interface IAzureTableStorageClient
+    {
+        Task<ITableEntity> RetrieveAsync<TElement>(string partitionKey, string rowKey, string tableName, string storageAccountConnectionString) where TElement : ITableEntity;
+        Task<ITableEntity> InsertOrMergeAsync(ITableEntity entity, string tableName, string storageAccountConnectionString);
+        Task<ITableEntity> DeleteAsync(ITableEntity entity, string tableName, string storageAccountConnectionString);
+        Task<CloudTable> GetCloudTable(string tableName, string storageAccountConnectionString);
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IAzureTableStorageClient.cs:11]: Kindly give more meaningful names to method like RetrieveFromTableStorageAsync, InsertOrMergeIntoTableStorageAsync,  DeleteFromTableStorageAsync, GetAzureTable respectively,

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceCacheTest.cs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        public async Task WhenInsertOrMergeFssCacheDetailReturnsTrue()
+        {            
+            A.CallTo(() => fakeAzureStorageService.GetStorageAccountConnectionString(A<string>.Ignored, A<string>.Ignored)).Returns(GetStorageAccountConnectionStringAndContainerName().Item1);
+            A.CallTo(() => fakeAzureTableStorageClient.InsertOrMergeAsync(A<TableEntity>.Ignored, A<string>.Ignored, A<string>.Ignored)).Returns(GetResponseCache());
+            await fileShareServiceCache.InsertOrMergeFssCacheDetail(GetResponseCache());
+
+            Assert.IsNotNull(true);
+        }
+    }
+}
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceCacheTest.cs:174]: Kindly change test method names to

WhenGetNonCacheProductDataForFssIsCalled_ThenReturnProductNotFound

WhenGetNonCacheProductDataForFssIsCalled_ThenReturnProductFound

WhenCancellationRequestedInGetNonCacheProductDataForFss_ThenThrowOperationCanceledException

WhenInsertOrMergeFssCacheDetail_ThenReturnTrue

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceCacheTest.cs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        public async Task WhenGetNonCacheProductDataForFssReturnsProductListNotFound()
+        {
+            var cachingResponse = new FssResponseCache() { };
+            A.CallTo(() => fakeAzureStorageService.GetStorageAccountConnectionString(A<string>.Ignored, A<string>.Ignored)).Returns(GetStorageAccountConnectionStringAndContainerName().Item1);
+            A.CallTo(() => fakeAzureTableStorageClient.RetrieveAsync<FssResponseCache>(A<string>.Ignored, A<string>.Ignored, A<string>.Ignored, A<string>.Ignored)).Returns(cachingResponse);
+
+            var response = await fileShareServiceCache.GetNonCacheProductDataForFss(GetProductdetails(), null, string.Empty, GetScsResponseQueueMessage(), null, CancellationToken.None);
+
+            Assert.IsNotNull(response);
+            Assert.IsInstanceOf(typeof(List<Products>), response);
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceCacheTest.cs:151]: Kindly assert count in first 2 test cases
Assert.AreEqual(0, response.Count);
Assert.AreEqual(1, response.Count);

and remove Assert.IsInstanceOf(typeof(List<Products>), response); from first 2 test cases as it is not required.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IAzureTableStorageClient.cs (hunk 2 above)
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IAzureTableStorageClient.cs]: done

**Review by janvi13603 — APPROVED** [review]
Thanks for fixing review comments.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 5]
```diff
@@ -406,14 +426,35 @@ private async Task<bool> ProcessBatchFile(IEnumerable<string> uri, string downlo
...
+                    Response = JsonConvert.SerializeObject(entry)
+                };
+                logger.LogInformation(EventIds.FileShareServiceSearchResponseStoreToCacheStart.ToEventId(), "File share service search response insert/merge request in azure table for cache started for Product/CellName:{ProductName}, EditionNumber:{EditionNumber} and UpdateNumber:{UpdateNumber} with FSS BatchId:{FssBatchId}. BatchId:{batchId} and _X-Correlation-ID:{CorrelationId}", productName, editionNumber, updateNumber, entry.BatchId, queueMessage.BatchId, queueMessage.CorrelationId);
+                await fileShareServiceCache.InsertOrMergeFssCacheDetail(fssSearchResponseCache);
+                logger.LogInformation(EventIds.FileShareServiceSearchResponseStoreToCacheCompleted.ToEventId(), "File share service search response insert/merge request in azure table for cache completed for Product/CellName:{ProductName}, EditionNumber:{EditionNumber} and UpdateNumber:{UpdateNumber} with FSS BatchId:{FssBatchId}. BatchId:{batchId} and _X-Correlation-ID:{CorrelationId}", productName, editionNumber, updateNumber, entry.BatchId, queueMessage.BatchId, queueMessage.CorrelationId);
```
Comment (barrie-cooper) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:444]: There is no test coverage for this new code.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 6]
```diff
@@ -406,14 +426,35 @@ private async Task<bool> ProcessBatchFile(IEnumerable<string> uri, string downlo
...
+            fileSystemHelper.CreateFileCopy(path, new MemoryStream(bytes));
+            if (!entry.IgnoreCache && fssCacheConfiguration.Value.IsFssCacheEnabled)
             {
-                fileSystemHelper.CreateFileCopy(path, stream);
+                logger.LogInformation(EventIds.FileShareServiceUploadENCFilesToCacheStart.ToEventId(), "File share service upload ENC file request to cache blob container started for Container:{Container}, with FileName: {FileName}. ESS BatchId:{batchId} and _X-Correlation-ID:{CorrelationId}", entry.BatchId, fileName, queueMessage.BatchId, queueMessage.CorrelationId);
+                await fileShareServiceCache.CopyFileToBlob(new MemoryStream(bytes), fileName, entry.BatchId);
+                logger.LogInformation(EventIds.FileShareServiceUploadENCFilesToCacheCompleted.ToEventId(), "File share service upload ENC file request to cache blob container completed for Container:{Container}, with FileName: {FileName}. ESS BatchId:{batchId} and _X-Correlation-ID:{CorrelationId}", entry.BatchId, fileName, queueMessage.BatchId, queueMessage.CorrelationId);
```
Comment (barrie-cooper) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:457]: There is no test coverage for this new code.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceCacheTest.cs [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+            return new List<Products> {
+                            new Products {
+                                ProductName = "DE416050",
+                                EditionNumber = 2,
+                                UpdateNumbers = new List<int?> {0},
+                                FileSize = 400
+                            }
+                        };
+        }
+        #endregion
```
Comment (barrie-cooper) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceCacheTest.cs:67]: Not a fan of regions so can we lose these directives please:
```suggestion
        private List<Products> GetProductdetails()
        {
            return new List<Products> {
                            new Products {
                                ProductName = "DE416050",
                                EditionNumber = 2,
                                UpdateNumbers = new List<int?> {0},
                                FileSize = 400
                            }
                        };
        }
```

**Review by barrie-cooper — COMMENTED** [review]
A few comments, most of them can be resolved with a button click 😃 

Re the test coverage comments if you want to do those later then that's fine with me (I know this is the Sprint Goal).

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileSystemHelper.cs [hunk 8]
```diff
@@ -135,5 +137,24 @@ private static string GetLine(string filePath)
...
+            byte[] buffer = new byte[16 * 1024];
+            using (MemoryStream ms = new MemoryStream())
+            {
+                int read;
+                while ((read = input.Read(buffer, 0, buffer.Length)) > 0)
+                {
+                    ms.Write(buffer, 0, read);
+                }
+                return ms.ToArray();
+            }
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileSystemHelper.cs:152]: As this code change may have an impact on code functionality and we have to test it properly, we will cover that in miscellaneous task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IAzureTableStorageClient.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.Azure.Cosmos.Table;
+using System.Threading.Tasks;
+
+namespace UKHO.ExchangeSetService.Common.Helpers
+{
+    public interface IAzureTableStorageClient
+    {
+        Task<ITableEntity> RetrieveFromTableStorageAsync<TElement>(string partitionKey, string rowKey, string tableName, string storageAccountConnectionString) where TElement : ITableEntity;
+        Task<ITableEntity> InsertOrMergeIntoTableStorageAsync(ITableEntity entity, string tableName, string storageAccountConnectionString);
+        Task<ITableEntity> DeleteFromTableStorageAsync(ITableEntity entity, string tableName, string storageAccountConnectionString);
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IAzureTableStorageClient.cs]: We have removed from code now

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        public async Task CopyFileToBlob(Stream stream, string fileName, string batchId)
+        {
+            var storageConnectionString = azureStorageService.GetStorageAccountConnectionString(fssCacheConfiguration.Value.CacheStorageAccountName, fssCacheConfiguration.Value.CacheStorageAccountKey);
+            CloudBlockBlob cloudBlockBlob = await azureBlobStorageClient.GetCloudBlockBlob(fileName, storageConnectionString, batchId);
+            cloudBlockBlob.Properties.ContentType = CONTENT_TYPE;
+            if (!await cloudBlockBlob.ExistsAsync())
+            {
+                await cloudBlockBlob.UploadFromStreamAsync(stream);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:131]: We will do this test code coverage in a separate Miscellaneous Task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs (hunk 5 above)
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:444]: We will do this test code coverage in a separate Miscellaneous Task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs (hunk 6 above)
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:457]: We will do this test code coverage in a separate Miscellaneous Task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                }
+                internalProductItemNotFound = new Products
+                {
+                    Cancellation = item.Cancellation,
+                    Dates = item.Dates,
+                    EditionNumber = item.EditionNumber,
+                    FileSize = item.FileSize,
+                    ProductName = item.ProductName
+                };
+                internalProductItemNotFound.UpdateNumbers = new List<int?>();
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:67]: As this code change may have an impact on code functionality and we have to test it properly, we will cover that in miscellaneous task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                    internalProductsNotFound.Add(internalProductItemNotFound);
+                }
+            }
+
+            return internalProductsNotFound;
+        }
+
+        private async Task<BatchDetail> CheckIfCacheProductsExistsInBlob(string exchangeSetRootPath, SalesCatalogueServiceResponseQueueMessage queueMessage, Products item, List<int?> updateNumbers, int? itemUpdateNumber, string storageConnectionString, FssSearchResponseCache cacheInfo)
+        {
+            BatchDetail internalBatchDetail;
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:99]: As this code change may have an impact on code functionality and we have to test it properly, we will cover that in miscellaneous task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            this.logger = logger;
+            this.azureStorageService = azureStorageService;
+            this.fssCacheConfiguration = fssCacheConfiguration;
+            this.fileSystemHelper = fileSystemHelper;
+        }
+        public async Task<List<Products>> GetNonCacheProductDataForFss(List<Products> products, SearchBatchResponse internalSearchBatchResponse, string exchangeSetRootPath, SalesCatalogueServiceResponseQueueMessage queueMessage, CancellationTokenSource cancellationTokenSource, CancellationToken cancellationToken)
+        {
+            var internalProductsNotFound = new List<Products>();
+            Products internalProductItemNotFound;
+            BatchDetail internalBatchDetail;
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:49]: As this code change may have an impact on code functionality and we have to test it properly, we will cover that in miscellaneous task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+                foreach (var itemUpdateNumber in item.UpdateNumbers)
+                {
+                    var compareProducts = $"{item.ProductName}|{item.EditionNumber.Value}|{itemUpdateNumber.Value}";
+                    if (!productList.Contains(compareProducts))
+                    {
+                        var storageConnectionString = azureStorageService.GetStorageAccountConnectionString(fssCacheConfiguration.Value.CacheStorageAccountName, fssCacheConfiguration.Value.CacheStorageAccountKey);
+                        var cacheInfo = (FssSearchResponseCache)await azureTableStorageClient.RetrieveFromTableStorageAsync<FssSearchResponseCache>(item.ProductName, item.EditionNumber + "|" + itemUpdateNumber.Value, fssCacheConfiguration.Value.FssSearchCacheTableName, storageConnectionString);
+                        if (cacheInfo != null && !string.IsNullOrEmpty(cacheInfo.Response))
+                        {
+                            internalBatchDetail = await CheckIfCacheProductsExistsInBlob(exchangeSetRootPath, queueMessage, item, updateNumbers, itemUpdateNumber, storageConnectionString, cacheInfo);
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:78]: As this code change may have an impact on code functionality and we have to test it properly, we will cover that in miscellaneous task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+            }
+
+            return internalProductsNotFound;
+        }
+
+        private async Task<BatchDetail> CheckIfCacheProductsExistsInBlob(string exchangeSetRootPath, SalesCatalogueServiceResponseQueueMessage queueMessage, Products item, List<int?> updateNumbers, int? itemUpdateNumber, string storageConnectionString, FssSearchResponseCache cacheInfo)
+        {
+            BatchDetail internalBatchDetail;
+            logger.LogInformation(EventIds.FileShareServiceSearchENCFilesFromCacheStart.ToEventId(), "File share service search request from cache started for Product/CellName:{ProductName}, EditionNumber:{EditionNumber} and UpdateNumber:{UpdateNumber}. BatchId:{batchId} and _X-Correlation-ID:{CorrelationId}", item.ProductName, item.EditionNumber, itemUpdateNumber, queueMessage.BatchId, queueMessage.CorrelationId);
+            internalBatchDetail = JsonConvert.DeserializeObject<BatchDetail>(cacheInfo.Response);
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:101]: As this code change may have an impact on code functionality and we have to test it properly, we will cover that in miscellaneous task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IFileShareServiceCache.cs [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+using System.Threading.Tasks;
+using UKHO.ExchangeSetService.Common.Models.AzureTableEntities;
+using UKHO.ExchangeSetService.Common.Models.FileShareService.Response;
+using UKHO.ExchangeSetService.Common.Models.SalesCatalogue;
+
+namespace UKHO.ExchangeSetService.Common.Helpers
+{
+    public interface IFileShareServiceCache
+    {
+        Task<List<Products>> GetNonCacheProductDataForFss(List<Products> products, SearchBatchResponse internalSearchBatchResponse, string exchangeSetRootPath, SalesCatalogueServiceResponseQueueMessage queueMessage, CancellationTokenSource cancellationTokenSource, CancellationToken cancellationToken);
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IFileShareServiceCache.cs]: As this code change may have an impact on code functionality and we have to test it properly, we will cover that in miscellaneous task on Monday

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (barrie-cooper) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs:101]: Will this not be covered by existing unit tests?

**Review by barrie-cooper — APPROVED** [review]
Thanks for changes.  Still some unit test coverage to improve but happy to have that as a separate task to follow.

**Pull request merged**